File: tutorial_ii.jl
#! tutorial ii

using Random
using CSV
using DataFrames
using Distributions
using LinearAlgebra
using Statistics
using Optim

Random.seed!(243587)

# load the data
data = CSV.read("data_ii.csv", DataFrame)

const X = Matrix(data[!, [:x1, :x2]])
const y = Vector(data[!, :y])

# draws of normal random numbers
const M = 10
const u = rand(Normal(0, sqrt(1.5)), length(y), M)

# define auxiliary model estimator
function ols(x, y; add_constant=true)
    if add_constant == true
        n = size(x, 1)
        x = hcat(ones(n), x)
    end
    x \ y # this is the same as (x'*x)^{-1} * x'*y, but faster and more accurate
end

# estimate aux model with observed data
const β̂ = ols(X, y)

# define structural model
function G(X, u, θ)
    n = size(X, 1)
    exp.(hcat(ones(n), X) * θ) + u
end

# define criterion for ii
wald(β̃; β̂=β̂) = sum((β̂ - β̃).^2)

# define objective function
function obj(θ; X=X, u=u, β̂=β̂)
    ỹ = [G(X, u[:, m], θ) for m in 1:M] # this simulates the data
    β̃ = mean(ols(X, ỹ[m]) for m in 1:M) # this estimates aux model on simulated data, and then averages parameters
    return wald(β̃; β̂=β̂)
end

# optimization of criterion
θ0 = ols(X, y) # initial vector of parameters
opt = optimize(θ -> obj(θ; X=X, u=u, β̂=β̂), θ0, NelderMead())
opt.minimizer

# the true values are
#=
 0.9279139553540544
 0.2408481106524355
 0.4354059904502885
=#

File: tutorial_ii.R
#! tutorial ii

# load required packages
library(methods)
library(utils)
library(stats)
library(optimx)

set.seed(238476)

# load the data
data <- read.csv("data_ii.csv")

X <- as.matrix(data[, c("x1", "x2")])
y <- as.matrix(data["y"])

# draws of normal random numbers
M <- 10
u <- matrix(rnorm(length(y)*M, mean = 0, sd = sqrt(1.5)), ncol=10)

# define auxiliary model estimator
ols <- function(X, y, add.constant=TRUE){
  if (add.constant==TRUE){
    n <- length(y)
    X <- cbind(rep(1, n), X)
  }
  solve(t(X) %*% X) %*% t(X) %*% y
}

# estimate aux model with observed data
beta.hat <- ols(X, y)

# define structural model
G <- function(X, u, theta){
  n <- dim(X)[2]
  exp(cbind(rep(1, n), X) %*% theta) + u
}

# define criterion for ii
wald <- function(beta.tilde, beta.hat){
  sum((beta.tilde - beta.hat)^2)
}

# define objective function
obj <- function(theta, X, u, beta.hat){
  y.tilde <- apply(u, 2, function(u) G(X, u, theta)) # this simulates the data
  beta.tilde <- apply(y.tilde, 2, function(y.tilde) ols(X, y.tilde)) # this estimates aux model on simulated data
  beta.tilde <- apply(beta.tilde, 1, mean) # this averages aux model estimates
  wald(beta.tilde, beta.hat)
}

# optimization of criterion
theta.0 <- as.vector(ols(X, y))
opt <- optimx(theta.0, function(theta) obj(theta, X, u, beta.hat), method="Nelder-Mead")
opt

# the true values are
# 0.9279139553540544
# 0.2408481106524355
# 0.4354059904502885
